core/trino-main/src/main/java/io/trino/connector/FileCatalogStore.java

```diff
@@ -71,16 +85,63 @@ public FileCatalogStore(StaticCatalogManagerConfig config)
             String connectorName = properties.remove("connector.name");
             checkState(connectorName != null, "Catalog configuration %s does not contain 'connector.name'", file.getAbsoluteFile());
 
-            CatalogHandle catalogHandle = createRootCatalogHandle(catalogName, computeCatalogVersion(catalogName, connectorName, properties));
-            catalogProperties.add(new CatalogProperties(catalogHandle, connectorName, ImmutableMap.copyOf(properties)));
+            catalogs.put(catalogName, createCatalogProperties(catalogName, connectorName, ImmutableMap.copyOf(properties)));
         }
-        this.catalogs = catalogProperties.build();
     }
 
     @Override
     public Collection<CatalogProperties> getCatalogs()
     {
-        return catalogs;
+        return ImmutableList.copyOf(catalogs.values());
+    }
+
+    @Override
+    public CatalogProperties createCatalogProperties(String catalogName, String connectorName, Map<String, String> properties)
+    {
+        if (readOnly) {
+            throw new TrinoException(NOT_SUPPORTED, "Catalog store is read only");
+        }
+        return new CatalogProperties(
+                createRootCatalogHandle(catalogName, computeCatalogVersion(catalogName, connectorName, properties)),
+                connectorName,
+                ImmutableMap.copyOf(properties));
+    }
+
+    @Override
+    public void addCatalog(CatalogProperties catalogProperties)
+    {
+        if (readOnly) {
+            throw new TrinoException(NOT_SUPPORTED, "Catalog store is read only");
+        }
+        File file = toFile(catalogProperties.getCatalogHandle().getCatalogName());
+        Properties properties = new Properties();
+        properties.setProperty("connector.name", catalogProperties.getConnectorName());
+        properties.putAll(catalogProperties.getProperties());
+
+        try (FileOutputStream fileWriter = new FileOutputStream(file)) {
```

Comment:
This can leave an empty or partially written file if we crash in the middle. We can make it safer by writing to a temporary file and renaming, and might want to fsync it as well:
```java
File temporary = new File(file.getPath() + ".tmp");
try (FileOutputStream writer = new FileOutputStream(temporary)) {
    properties.store(writer, null);
    temporary.flush();
    temporary.getFD().sync();
}
Files.move(temporary.toPath(), file.toPath(), REPLACE_EXISTING);